developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown

```diff
@@ -0,0 +1,76 @@
+---
+header-id: filtering-and-searching-uad-marked-entities
+---
+
+# Filtering and Searching UAD-Marked Entities
+
+[TOC levels=1-4]
+
+When you 
+
+![Figure 1: The UAD UI has filtering and searching elements you can leverage in
+custom entities.](../../images/uad-filter-search.png)
+
+To add filtering and searching for your custom entities, implement three
+methods in the `UADDisplay` class (in your application's `-uad` module):
+
+## Filtering
+
+The `isSiteScoped` method returns a boolean based on whether or not the
+entities can be associated with a particular site. This is used to determine
+which filter they will be associated with ("instance", "personal-site", or
+"regular-sites").
+
+<!-- I want to show some realistic dummy code for third party devs. Is this fine?-->
+<!--Straight out of BaseModelUADDisaply: -->
+
+```java
+@Override
+public boolean isSiteScoped() {
+    Class<FooEntry> typeClass = FooEntry.class;
+
+    if (GroupedModel.class.isAssignableFrom(typeClass)) {
+        return true;
+    }
+
+    return false;
+}
+```
+
+## Search
+
+Implement the `search` and `searchCount` methods to enable search in the UAD
+interface:
+
+<!--Is this realistic dummy code for third party devs? -->
+<!-- Straight out of DummyUADDisplay.java -->
```

Comment:
It is ok, but we should encourage the devs to actually implement some sort of real search (with filtering) if possible.  The `DummyUADDisplay` was designed to test other capabilities, not to template what a search implementation should actually look like.

Reply:
Do you mean, code to call the entity's indexer.search?

Create a searchContext (`SearchContextFactory.getInstance`), get the Indexer (`IndexerRegistryUtil.getIndexer(Entry.class)`, then execute the indexer.search?

There's no example of this in any portal app that I can see, so I could use some example code if possible.
@drewbrokke

Reply:
Yes, that kind of example would work well.  We went for trying to cover as many cases as possible inside of the core product, and if a customer follows our ServiceBuilder-based instructions they'll end up with that base case code OOTB. But if they want to implement this method themselves, it's likely the'll want to use the indexer if their entity has it.  Ultimately they can use any approach they want so long as the `search` method returns items that are related to the user and match the given search parameters. What that means exactly is up to each app.

Also, `SearchContext` doesn't require a request, you may be thinking of `SearchContext.getInstance`.  However, there's a different `SearchContext.getInstance` that doesn't require a request but instead asks for individual parameters (see https://github.com/liferay/liferay-portal/blob/22502f6df640d5b33f4d507354e327174f34159e/portal-kernel/src/com/liferay/portal/kernel/search/SearchContextFactory.java#L120).

Reply:
@drewbrokke 
How does this look for example code? I don't know where the Layout or the groupId would come from in this context, maybe you can help me with that:

```java
SearchContext searchContext = SearchContextFactory.getInstance(
    new long[0], new String[0], null, PortalUtil.getDefaultCompanyId(),
    keywords, NEED LAYOUT,
    LocaleThreadLocal.getSiteDefaultLocale(), NEED GROUPID,
    TimeZoneThreadLocal.getDefaultTimeZone(), userId);

searchContext.setStart(start);
searchContext.setEnd(end);

Indexer indexer = IndexerRegistryUtil.getIndexer(FooEntry.class);

Hits hits = indexer.search(searchContext);

List<FooEntry> fooEntries = new ArrayList<FooEntry>();

    for (int i = 0; i < hits.getDocs().length; i++) {
            Document doc = hits.doc(i);

            long entryId = GetterUtil
            .getLong(doc.get(Field.ENTRY_CLASS_PK));

            Entry entry = null;

            try {
                    entry = FooEntryLocalService.getFooEntry(fooEntryId);
            } catch (PortalException pe) {
                    _log.error(pe.getLocalizedMessage());
            } catch (SystemException se) {
                    _log.error(se.getLocalizedMessage());
            }

            fooEntries.add(fooEntry);
    }
```

Reply:
@rbohl unfortunately we're at the edge of my authority, from here on out you get my best guesses ;) 

I don't know about `Layout` in this context, but for `groupId` you can just use 0, and then use `searchContext.setGroupIds(groupIds)` after the searchContext is created.

You'd ultimately be better off talking to one of the members of the search team to come up with a solid example.

Reply:
@rbohl on second look, I don't see any usages of `searchContext.getLayout`, so as far as sample code is concerned you're pretty safe to pass in pretty much anything for the `Layout` at this point.

Reply:
Thanks Drew.

@lipusz do you have any input about the example`indexer.search` call I've cobbled together in the above code snippet. I'll take Drew's suggestion about the groupIds, but haven't added it yet. The context is, we're wanting to show sample code for writing the `search` method of the UADDisplay interface.

Reply:
@rbohl 

You can just use 
```java
SearchContext searchContext = new SearchContext();
```

and then set what you need. You don't need a `Layout` here.

Reply:
One question regarding the API:
> `@param  userId the primary key of the user`

So `userId` refers to the `User` the invoker wants to retrieve the data associated with or it's the `userId` of the invoker User?

Reply:
I'm asking because `searchContext.setUser(long userId)` is used to include the permission "pre-filters" in the query to filter the results by VIEW permission so here you'd need to `userId` of the invoker `User`.

Otherwise you need to post-filter the results.

Reply:
Good question. I think this is the User ID not of the invoker, but of the User whose data is being sanitized. @drewbrokke ?

Reply:
@lipusz and @drewbrokke 

Here's what I have right now, for populating the searchContext:

```java
    SearchContext searchContext = new SearchContext();

    searchContext.setStart(start);
    searchContext.setEnd(end);
    searchContext.setGroupIds(groupIds);
    searchContext.setKeywords(keywords);
    searchContext.setUserId(???);

    Indexer indexer = IndexerRegistryUtil.getIndexer(FooEntry.class);

    Hits hits = indexer.search(searchContext);
```

As you can see, not sure what to pass as the userId, since the userId we have here is the ID of the user whose data we're searching for. If we don't pass anything, Tibor mentioned we'll have to do post-filtering. I'm a bit stuck as to how I proceed here. We're in the Control Panel, using admin only features. Not sure if that matters.
:thinking:

Reply:
Does this get me anywhere? Idea stolen from the powwow portlet's ZoomAPICallbackServlet class :man_shrugging: 

```java
    BooleanQuery booleanQuery = BooleanQueryFactoryUtil.create(
        searchContext);

    booleanQuery.addExactTerm("userId", userId);

    BooleanClause booleanClause = BooleanClauseFactoryUtil.create(
        booleanQuery, BooleanClauseOccur.MUST.getName());

    searchContext.setBooleanClauses(new BooleanClause[] {booleanClause});
```

Reply:
@drewbrokke I'm gonna take this directly to the search team via the #help-search channel in slack (or #t-search; so many choices). That way we don't just overload Tibor, but give the whole team a chance to respond as they have time.